code/modules/holomap/holomap.dm

```diff
@@ -0,0 +1,491 @@
+				All the following logic is to determine what icon to use.
+			*/
+			//this is what all the logic is for, setting this variable
+			var/icon_state_to_use
+			//as well as this since it modifies the wall sprite
+			var/dir_to_use = SOUTH
+
+			//the following variables are set conditionally on the state of the turf
+			//so we can render the right sprite for this turf
+			var/neighbor_space = FALSE
+			var/list/neighbors_alien = list()
+			var/list/neighbors_solid = list()
+
+	//--------------{{{ DISCOVERY PHASE }}}------------------------------
+
+			for(var/j in GLOB.diagonals)
+				var/turf/found_s_turf = get_step(T, j)
+				var/area/found_s_area = found_s_turf.loc
+				if(found_s_area == /area/space || found_s_area == /area/space/nearstation || isspaceturf(found_s_turf))
+					neighbor_space = TRUE
+					break
+
+			for(var/i in GLOB.cardinal)
+				var/turf/found_turf = get_step(T, i)
+				var/area/found_area = found_turf.loc
+				var/found_window = ((locate(/obj/structure/window/full) in found_turf) || (locate(/obj/effect/spawner/window) in found_turf)) ? TRUE : FALSE
+
+				if(neighbor_space || found_area == /area/space || found_area == /area/space/nearstation || isspaceturf(found_turf))
+					neighbor_space = TRUE
+					break
+
+				if(found_area != A || locate(/obj/machinery/door/airlock) in found_turf)
+					neighbors_alien += i
+					continue
+
+				if(found_window ||found_turf.density)
+					neighbors_solid += i
+
+	//-----------------{{{ RENDERING PHASE }}}----------------------------
+
+			if(neighbor_space || length(neighbors_solid) >= 3)
+				if( W || T.density)
+					icon_state_to_use = "full"
+				else if(isfloorturf(T))
+					icon_state_to_use = "floor"
+
+			//--> (not next to space; not surrounded; less than 3 solid neighbors)
+			else if(W || T.density)
+				if(length(neighbors_alien) == 1)
+					icon_state_to_use = "wall"
+					if(length(neighbors_solid) == 1)
+						if((neighbors_alien[1] == NORTH && neighbors_solid[1] == SOUTH) || (neighbors_alien[1] == WEST && neighbors_solid[1] == EAST))
+							dir_to_use = NORTHWEST
+						else if(neighbors_alien[1] == SOUTH && neighbors_solid[1] == NORTH)
+							dir_to_use = SOUTHWEST
+						else if(neighbors_alien[1] == EAST && neighbors_solid[1] == WEST)
+							dir_to_use = NORTHEAST
+						else
+							dir_to_use = neighbors_alien[1]
+					else
+						dir_to_use = neighbors_alien[1]
+				else if(length(neighbors_alien) == 2)
+					icon_state_to_use = "wall"
+					if((neighbors_alien[1] == NORTH && neighbors_alien[2] == SOUTH))
+						icon_state_to_use = "full"
+					else if((neighbors_alien[1] == NORTH && neighbors_alien[2] == EAST))
+						dir_to_use = NORTHEAST
+					else if(neighbors_alien[1] == NORTH && neighbors_alien[2] == WEST)
+						dir_to_use = NORTHWEST
+					else if(neighbors_alien[1] == SOUTH && neighbors_alien[2] == EAST)
+						dir_to_use = SOUTHEAST
+					else if(neighbors_alien[1] == SOUTH && neighbors_alien[2] == WEST)
+						dir_to_use = SOUTHWEST
```

Comment:
Yo, can you explain a little about the repeat checks against the same thing here?

Reply:
none of them are repeats, we know there are two turfs adjacent that are "alien". In every scenario we render this turf differently ---> if they are north and south or if they are north and east, etc etc

because of how we check and store the data the second alien cannot have a smaller dir value than the first, eliminating alot of checks. I just go through all the possible scenarios here.